Katana/SideEffect.swift

```diff
@@ -11,66 +11,141 @@ import Hydra
    - returns: a promise that is resolved when the store finishes handling the dispatched item
    */
   @discardableResult
   func anyDispatch(_ dispatchable: Dispatchable) -> Promise<Any>
 
   /**
-  Dispatches a `ReturningSideEffect`. It is a partially erased version the `dispatch<T: SideEffect>(_:)` where only the return value is typed.
+   Dispatches a `Dispatchable`.
 
-  - parameter dispatchable: the side effect to dispatch
+   - parameter dispatchable: the item to dispatch
+   - returns: a promise that is resolved when the store finishes handling the dispatched item
+   */
+  @discardableResult
+  func dispatch<T: Dispatchable>(_ dispatchable: T) -> Promise<Any>
+
+  /**
+  Dispatches a `AnySideEffect`.
+
+  - parameter dispatchable: the `AnySideEffect` to dispatch
   - returns: a promise that is resolved with the value returned by the side effect
              when the store finishes handling the dispatched item
   */
   @discardableResult
-  func dispatch<T: ReturningSideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue>
-  
+  func dispatch<T: AnySideEffect>(_ dispatchable: T) -> Promise<Void>
+
   /**
-   Dispatches a `AnyStateUpdater`. By considering that all the `StateUpdater`s are also `AnyStateUpdater`s this dispatch
-   allows to dispatch all the possible `StateUpdaters` an app or a lib can implement.
-   
+   Dispatches a `AnyStateUpdater`.
+
    - parameter dispatchable: the side effect to dispatch
    - returns: a promise that is resolved when the store finishes updating the state
    */
   @discardableResult
   func dispatch<T: AnyStateUpdater>(_ dispatchable: T) -> Promise<Void>
+
+  /**
+  Dispatches a `ReturningSideEffect`
+
+  - parameter dispatchable: the `ReturningSideEffect` to dispatch
+  - returns: a promise that is resolved with the value returned by the side effect
+             when the store finishes handling the dispatched item
+  */
+  @discardableResult
+  func dispatch<T: ReturningSideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue>
 }
 
 public extension AnySideEffectContext {
+  @discardableResult
+  func dispatch<T: Dispatchable>(_ dispatchable: T) -> Promise<Any> {
+    return self.anyDispatch(dispatchable)
+  }
+
   /// Default implementation of the `dispatch<T: AnyStateUpdater>`
+  @discardableResult
   func dispatch<T: AnyStateUpdater>(_ dispatchable: T) -> Promise<Void> {
     return self.anyDispatch(dispatchable).void
   }
 
+  /// Default implementation of the `dispatch<T: AnySideEffect>`
+  @discardableResult
+  func dispatch<T: AnySideEffect>(_ dispatchable: T) -> Promise<Void> {
+    return self.anyDispatch(dispatchable).void
+  }
+
   /// Default implementation of the `dispatch<T: ReturningSideEffect>`
+  @discardableResult
   func dispatch<T: ReturningSideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue> {
     return self.anyDispatch(dispatchable).then { $0 as! T.ReturnValue }
   }
 }
 
+/// Extension that contains some helper method
+public extension AnySideEffectContext {
```

Comment:
I think we can remove this extension. If we are moving toward libs with few concepts that can be composed together, the methods in this extensions are just helpers that put together `await` and `dispatch`. Thus, they are increasing the API surface without providing a real gain for the lib itself.
Also, we end up having multiple ways to perform the same operation.